File: packages/backend/src/conversation/failure-facts.ts
import type { Block, ProviderFailureFacts } from '@demicodes/core'
import type { ProviderFailureReader } from '@demicodes/provider'
import type { ProviderAssembly } from '../llm/assembly'

/** The facts of a transcript's error blocks, by block id, as the browser gets them. */
export type FailureFacts = Record<string, ProviderFailureFacts>

/** Reads the failure facts of a list of blocks; the one function both delivery paths call. */
export type FailureFactsReader = (blocks: readonly Block[]) => Promise<FailureFacts>

/**
 * The failure facts of a list of blocks (`docs/demi-next/backend.md` § Failure
 * facts): each error block with a record is read by the provider named in its
 * model selection. A block whose provider configuration is gone, or whose
 * provider reads nothing, yields no entry.
 */
export function failureFactsReader(assembly: ProviderAssembly): FailureFactsReader {
  return async (blocks) => {
    const failures: FailureFacts = {}
    const readers = new Map<string, Promise<ProviderFailureReader | null>>()
    for (const block of blocks) {
      if (block.type !== 'error' || block.diagnostics?.upstream === undefined)
        continue
      const providerId = block.model.providerId
      if (!readers.has(providerId))
        readers.set(providerId, readerOf(assembly, providerId))
      const read = await readers.get(providerId)!
      if (read)
        failures[block.id] = read(block.diagnostics, block.createdAt)
    }
    return failures
  }
}

/**
 * The reader of a configured provider. Facts only add to a record the browser
 * gets anyway, so a configuration that cannot be read yields no reader rather
 * than holding back the transcript.
 */
async function readerOf(
  assembly: ProviderAssembly,
  providerId: string
): Promise<ProviderFailureReader | null> {
  try {
    return (await assembly.providerFor(providerId))?.provider.readFailure ?? null
  } catch {
    return null
  }
}

File: crates/backend/src/conversation/failure_facts.rs
//! The failure facts of a transcript's error blocks (`backend.md` § Failure
//! facts): each error block with a failure record is read by the provider
//! its model selection names, and the facts travel beside the blocks, never
//! stored. The conversation socket's transcript frames and the history
//! route both read them here.

use std::collections::HashMap;
use std::sync::Arc;

use demi_agent_protocol::Failures;
use demi_core::Block;
use demi_provider::Provider;
use demi_web_api::ids::ProviderId;

use crate::llm::assembly::ProviderAssembly;

/// The facts of the error blocks among `blocks`, by block id; none when no
/// block yields one. A block whose provider entry is gone, or cannot be
/// read, shows without facts.
pub(crate) async fn failure_facts(assembly: &ProviderAssembly, blocks: &[Block]) -> Option<Failures> {
    let mut failures = Failures::new();
    let mut readers: HashMap<String, Option<Arc<dyn Provider>>> = HashMap::new();
    for block in blocks {
        let Block::Error(error) = block else {
            continue;
        };
        let Some(diagnostics) = error.diagnostics.as_ref().filter(|diagnostics| diagnostics.upstream.is_some()) else {
            continue;
        };
        let provider = &error.model.provider_id;
        if !readers.contains_key(provider) {
            let reader = reader(assembly, provider).await;
            readers.insert(provider.clone(), reader);
        }
        if let Some(Some(reader)) = readers.get(provider) {
            failures.insert(error.id.clone(), reader.read_failure(diagnostics, error.created_at));
        }
    }
    (!failures.is_empty()).then_some(failures)
}

/// The provider of the entry `provider`. Facts only add to what the browser
/// receives anyway, so an entry that is gone or cannot be read yields no
/// reader rather than holding back the transcript.
async fn reader(assembly: &ProviderAssembly, provider: &str) -> Option<Arc<dyn Provider>> {
    let id = ProviderId::try_from(provider).ok()?;
    let entry = match assembly.vault().entry(id).await {
        Ok(entry) => entry?,
        Err(error) => {
            tracing::warn!(provider, error = &error as &dyn std::error::Error, "a failure's provider entry was not read");
            return None;
        }
    };
    match assembly.provider_for(&entry).await {
        Ok(provider) => Some(provider),
        Err(error) => {
            tracing::warn!(provider = %entry.id, error = &error as &dyn std::error::Error, "a failure's provider was not built");
            None
        }
    }
}
